src/state/products.js

```diff
@@ -39,6 +39,39 @@ export const extractedProductShape = pt.shape({
   date: pt.string.isRequired,
 });
 
+/**
+ * Validate the contents of the given extracted product. This is used in lieu
+ * of proptypes since proptype checks do not work in a production build.
```

Comment:
Why doesn't this work in production? Why is it not feasible to make proptype checks work in production?

Does this also affect the prop types in the browserAction? Or could we use this method of validation where we could not use `checkPropTypes`? For example:

```javascript
export default class BrowserActionApp extends React.Component {
  static propTypes = {
    // Direct props
    extractedProduct: extractedProductShape, // Product detected on the current page, if any

    // State props
    products: pt.arrayOf(productShape).isRequired,

    // Dispatch props
    loadStateFromStorage: pt.func.isRequired,
  }

  static defaultProps = {
    extractedProduct: null,
  }
//...
```

Reply:
> Why doesn't this work in production?

Proptypes do not do anything in a production build. In production builds, Webpack sets `process.env.NODE_ENV` to `"production"`, which is the Node syntax for specifying the `NODE_ENV` environment variable. `NODE_ENV` is often used to signify whether JavaScript is running in a development or production environment.

Lots of frontend JavaScript adopted this convention to gate code that should be skipped in production for performance or other reasons. React considers proptype validation to be too slow to run on production websites, so they wrap most of the proptype validation code in `if (process.env.NODE_ENV !== "production")`. Since Webpack is setting this value to `"production"`, the code doesn't run, saving time by skipping the validation step (it's assumed that you'll notice proptype errors during testing in development mode and fix them).

(In fact, Webpack goes a step further; the minifying step detects that the if statement will never run and actually removes the code in it , so in a production build of Webpack, the proptype code doesn't even exist. This is called "Tree Shaking". React does this with a lot of its code, which is why a production build of React is much smaller than a development build. There's literally less code.)

> Why is it not feasible to make proptype checks work in production?

There's no way to make the `process.env.NODE_ENV` variable set to something besides `"production"` for only certain files, so if we change it our built zip file will become way bigger and run a lot of development-only React code. Due to how the checks in the PropTypes code are written, there's no easy way of removing it without maintaining our own forked copy. The only maintainable option left is to use a different type of validation besides PropTypes.

> Does this also affect the prop types in the browserAction?

Yes, the prop types in the browser action do not run in a production build, but this is expected behavior and we don't need to change it. Proptype errors should be logged in development mode, and we should be checking that log and/or (in the future) have our automated tests assert that no messages get logged when the components are rendered.

> Or could we use this method of validation where we could not use `checkPropTypes`?

If you mean "can we use proptypes where we can't use checkproptypes", the answer is no; whether run via `checkPropTypes` or added to a React component, proptypes are always optimized away in a production build.

Reply:
Webpack is the bit that removes it since it's performing the build, and yes, it does.

To me, the split between devDependencies and dependencies is that dependencies is for things that are used in your app code, even if it gets optimized away. So prop-types stays in even though it gets optimized out.